Fixed away_mode for Ecobee thermostat.

etsinko opened the pull request:
## Description:

Fix for away_mode for Ecobee thermostat. Now away mode is properly turned on using indefinite away hold, the correct value is returned for is_away_mode_on() and finally away mode is correctly disabled using turn_away_mode_off()

In addition two more bugs in ecobee platform have been fixed:

1. Ecobee platform had some weird rounding errors due to all temperature holds being cast to integers before being passed to the thermostat. i.e. if you set temperature hold as 73.6 degrees upper/ 35.2 degrees lower it was passed to ecobee as 74 degrees upper / 35 lower. (This works in conjuction with python-ecobee-api v 0.0.10 which has been updated in #9612)

2. If temperature hold was set while in auto mode an exception was thrown. This was due to set_temp_hold() method covering only two cases (Cooling / Heating) and not covering auto mode

**Related issue (if applicable):** **#9256**

**Pull request in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io) with documentation (if applicable):** **N/A**

## Example entry for `configuration.yaml` (if applicable): **N/A**
## Checklist:

If user exposed functionality or configuration variables are added/changed: **N/A**

If the code communicates with devices, web services, or third-party tools: **N/A**
If the code does not interact with devices:
  - [x] Local tests with `tox` run successfully. **Your PR cannot be merged unless tests pass**

[ex-requir]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L14
[ex-import]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L54

etsinko committed:
Fixed away_mode for Ecobee thermostat. Now away mode is properly turned on using indefinite away hold.

homeassistant commented:
Hi @etsinko, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;home-assistant#9559).

Once you do that we will be able to review and accept this pull request.

Thanks!

etsinko committed:
fixed lint warnings

MartinHjelmare commented:
Follow and fill in the PR template that was there when you opened the PR. Eg if there's an issue with background info, it should be linked to.

etsinko committed:
fixed lint warnings

etsinko commented:
Done,
no tests have been added because the original class didn't have any tests

arsaboo commented:
@etsinko Wondering if it also fixes https://github.com/home-assistant/home-assistant/issues/9256, in which case you can indicate it in the PR.

MartinHjelmare commented:
Isn't this related?
https://github.com/home-assistant/home-assistant/issues/9256

etsinko commented:
@arsaboo It doesn't really fix it. Interestingly enough hold_temp is set into the thermostat but is never used since b732174def5894708e90c7d8708e971410a43afd

etsinko commented:
@MartinHjelmare 
I think it is only related by 50% because #9256 refers to two problems:
1. hold_temp doesn't work
2. away_mode doesn't work

This PR fixes problem 2.

MartinHjelmare commented:
Evaluating this PR is very difficult if you're not very familiar with the device api (ecobee). Giving as much background info as possible will help the review. 50% related issue is related.

etsinko commented:
@MartinHjelmare I think you are right. I added it to the description.

dancwilliams commented:
@etsinko I have been running your `ecobee.py` component in my environment the past couple of days and it is operating as I would expect it to.  @chilicheech have you given this a look/run?

I use the following automatons in my environment to set holds:

```
- id: set_ecobee_mode_away
  alias: Set Main House Ecobee Mode Away
  trigger:
    platform: state
    entity_id: input_select.main_house_thermostat_mode
    to: away
  action:
    - service: climate.set_away_mode
      data_template:
        entity_id: climate.main_house
        away_mode: on

- id: set_ecobee_mode
  alias: Set Main House Ecobee Mode
  trigger:
    - platform: state
      entity_id: input_select.main_house_thermostat_mode
      to: home
    - platform: state
      entity_id: input_select.main_house_thermostat_mode
      to: sleep
    - platform: state
      entity_id: input_select.main_house_thermostat_mode
      to: smart1
  action:
    - service: climate.set_hold_mode
      data_template:
        entity_id: climate.main_house
        hold_mode: '{{ states.input_select.main_house_thermostat_mode.state }}'
```

I also tested setting a hold to `away` using the `climate.set_hold_mode` service and it set the mode to away without being permanent, which is what I would expect.

etsinko commented:
I have also found a couple more bugs in this platform and in python-ecobee-api which is a requirement for this platform.
I will update this PR with new changes along with unit tests as soon as python-ecobee-api (https://github.com/nkgilley/python-ecobee-api) is fixed (hopefully soon, I did a PR there too)

etsinko committed:
- now it is possible to use float values for ecobee temperature holds
- fixed a bug that caused an exception when temperature hold was set in away mode
- added unit tests for ecobee thermostat

etsinko committed:
- now it is possible to use float values for ecobee temperature holds
- fixed a bug that caused an exception when temperature hold was set in away mode
- added unit tests for ecobee thermostat

etsinko commented:
I fixed two more bugs in ecobee platform:
1. Ecobee platform had some weird rounding errors due to all temperature holds being cast to integers before being passed to the thermostat. i.e. if you set temperature hold as 73.6 degrees upper/ 35.2 degrees lower it was passed to ecobee as 74 degrees upper / 35 lower.
2. If temperature hold was set while in auto mode an exception was thrown. This was due to set_temp_hold() covering only two cases (Cooling / Heating) and not covering auto mode

Also unit tests for the platform were added.

etsinko committed:
fixed lint errors

etsinko committed:
fixed lint errors

pvizeli commented:
Good work
